drivers/gnss/gnss_quectel_lcx6g.c

```diff
@@ -0,0 +1,794 @@
+		break;
+	}
+
+	ret = gnss_nmea0183_snprintk(data->dynamic_request_buf, sizeof(data->dynamic_request_buf),
+				     "PAIR752,%u,%u", pps_mode, config->pps_pulse_width);
+	if (ret < 0) {
+		return ret;
+	}
+
+	data->dynamic_script_chat.request_size = ret;
+
+	ret = gnss_nmea0183_snprintk(data->dynamic_match_buf, sizeof(data->dynamic_match_buf),
+				     "PAIR001,752,0");
+	if (ret < 0) {
+		return ret;
+	}
+
+	data->dynamic_match.match_size = ret;
+
+	return modem_chat_run_script(&data->chat, &data->dynamic_script);
+}
+
+static int quectel_lcx6g_resume(const struct device *dev)
+{
+	struct quectel_lcx6g_data *data = dev->data;
+	int ret;
+
+	ret = modem_pipe_open(data->uart_pipe);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = modem_chat_attach(&data->chat, data->uart_pipe);
+	if (ret < 0) {
+		modem_pipe_close(data->uart_pipe);
+		return ret;
+	}
+
+	ret = modem_chat_run_script(&data->chat, &resume_script);
+	if (ret < 0) {
+		modem_pipe_close(data->uart_pipe);
+		return ret;
+	}
+
+	k_msleep(1000);
+
+	ret = quectel_lcx6g_configure_pps(dev);
+	if (ret < 0) {
+		modem_pipe_close(data->uart_pipe);
+	}
+
+	return ret;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int quectel_lcx6g_suspend(const struct device *dev)
+{
+	struct quectel_lcx6g_data *data = dev->data;
+	int ret;
+
+	ret = modem_chat_run_script_run(&data->chat, &suspend_script);
+	if (ret < 0) {
+		modem_pipe_close(data->uart_pipe);
+	}
+
+	return ret;
+}
+
+static int quectel_lcx6g_turn_off(const struct device *dev)
+{
+	struct quectel_lcx6g_data *data = dev->data;
+
+	return modem_pipe_close(data->uart_pipe);
+}
+
+static int quectel_lcx6g_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	struct quectel_lcx6g_data *data = dev->data;
+	k_spinlock_key_t key;
+	int ret = -ENOTSUP;
+
+	key = k_spin_lock(&data->lock);
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		ret = quectel_lcx6g_suspend(dev);
+		break;
+
+	case PM_DEVICE_ACTION_RESUME:
+		ret = quectel_lcx6g_resume(dev);
+		break;
+
+	case PM_DEVICE_ACTION_TURN_ON:
+		ret = 0;
+		break;
+
+	case PM_DEVICE_ACTION_TURN_OFF:
+		ret = quectel_lcx6g_turn_off(dev);
+		break;
+
+	default:
+		break;
+	}
+
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+static int quectel_lcx6g_set_fix_rate(const struct device *dev, uint32_t fix_interval_ms)
+{
+	struct quectel_lcx6g_data *data = dev->data;
+	k_spinlock_key_t key;
+	int ret;
+
+	if (fix_interval_ms < 100 || fix_interval_ms > 1000) {
+		return -EINVAL;
+	}
+
+	key = k_spin_lock(&data->lock);
+
+	ret = gnss_nmea0183_snprintk(data->dynamic_request_buf, sizeof(data->dynamic_request_buf),
+				     "PAIR050,%u", fix_interval_ms);
+	if (ret < 0) {
+		k_spin_unlock(&data->lock, key);
+		return ret;
+	}
+
+	data->dynamic_script_chat.request_size = ret;
+
+	ret = gnss_nmea0183_snprintk(data->dynamic_match_buf, sizeof(data->dynamic_match_buf),
+				     "PAIR001,050,0");
+	if (ret < 0) {
+		k_spin_unlock(&data->lock, key);
+		return ret;
```

Comment:
suggestion: usage of `goto` and a single unlock point is, in general, safer (harder to miss a spinlock locked).

Reply:
Good suggestion, I even found a missed unlock while updating the driver :)